python/src/iceberg/catalog/base.py

```diff
@@ -0,0 +1,223 @@
+            location: Location for the table. Optional Argument.
+            partition_spec: PartitionSpec for the table. Optional Argument.
+            properties: Table metadata that can be a string based dictionary. Optional Argument.
+
+        Returns:
+            Table: the replaced table instance with the updated state
+
+        Raises:
+            TableNotFoundError: If a table with the name does not exist
+        """
+
+    @abstractmethod
+    def create_namespace(self, namespace: Tuple[str, ...], properties: Optional[Dict[str, str]] = None) -> None:
+        """Create a namespace in the catalog.
+
+        Args:
+            namespace: The namespace to be created. Tuple of namespace levels. Ex: ('com','org','dept')
+            properties: A string dictionary of properties for the given namespace
+
+        Raises:
+            AlreadyExistsError: If a namespace with the name already exists in the namespace
+        """
+
+    @abstractmethod
+    def drop_namespace(self, namespace: Tuple[str, ...]) -> None:
+        """Drop a namespace.
+
+        Args:
+            namespace: The namespace to be dropped. Tuple of namespace levels. Ex: ('com','org','dept')
+
+        Raises:
+            NamespaceNotFoundError: If a namespace with the name does not exist in the namespace
+            NamespaceNotEmptyError: If the namespace is not empty
+        """
+
+    @abstractmethod
+    def list_tables(self, namespace: Optional[Tuple[str, ...]] = None) -> List[Tuple[Tuple[str, ...], str]]:
+        """List tables under the given namespace in the catalog.
+
+        If namespace not provided, will list all tables in the catalog.
+
+        Args:
+            namespace: the namespace to search. Tuple of namespace levels. Ex: ('com','org','dept')
+
+        Returns:
+            List[Tuple[str, str]]: list of tuple of table namespace and their names.
+
+        Raises:
+            NamespaceNotFoundError: If a namespace with the name does not exist in the namespace
+        """
+
+    @abstractmethod
+    def list_namespaces(self) -> List[Tuple[str, ...]]:
+        """List namespaces from the given namespace. If not given, list top-level namespaces from the catalog.
+
+        Returns:
+            List[Tuple[str, ...]]: a List of namespace, where each element is a Tuple of namespace levels. Ex: ('com','org','dept')
+        """
+
+    @abstractmethod
+    def load_namespace_metadata(self, namespace: Tuple[str, ...]) -> Dict[str, str]:
+        """Get metadata dictionary for a namespace.
+
+        Args:
+            namespace: A Tuple of namespace levels. Ex: ('com','org','dept')
+
+        Returns:
+            Dict[str, str]: a dictionary of properties for the given namespace
+
+        Raises:
+            NamespaceNotFoundError: If a namespace with the name does not exist in the namespace
+        """
+
+    @abstractmethod
+    def set_namespace_metadata(self, namespace: Tuple[str, ...], metadata: Dict[str, str]) -> None:
```

Comment:
This needs to be very clear about behavior. The verb `set` can be interpreted 2 ways: set the whole metadata, or set the given properties. In most parts of the API, we prefer to be very clear and use separate args or methods for setting properties or removing properties.

I think this should do the same. How about changing this to `update_namespace` or `update_namespace_metadata` and accepting both `updates: Dict[str, str]` to set and `removals: Set[str]` to remove?

Reply:
I like `update_namespace` accepting both `updates` and `removals`, will make the arguments optional as well.

Reply:
Fixed in a recent commit. Marking it resolved.